### ISSUE TITLE
Docs: [no-unnecessary-condition mention subtleties of primitives' assignability

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10970]
### Before You File a Documentation Request Please Confirm You Have Done The Following...

- [x] I have looked for existing [open or closed documentation requests](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Adocumentation) that match my proposal.
- [x] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Suggested Changes

I'd like to add a section that _briefly_ alerts users to the nuances around primitives being assignable to object types, causing sort-of-false-positives with no-unnecessary-condition's truthiness checks. 

I'd like to mention:

- "object types" are just "things with properties", _not_ objects (a superset of [no-empty-object-type](https://typescript-eslint.io/rules/no-empty-object-type/) shenanigans)
   ```ts
   function foo(x: { toString: Function }) {
      if (x) { // reported by no-unnecessary-condition as always-truthy
      }
   }

   foo(false); 
   foo(true);
   foo({});
   ```

- `object` can easily contain primitives at runtime, without typical obviously unsafe TS patterns.
   ```ts
   function foo(o: object) {
     if (o) { // reported by no-unnecessary-condition as always-truthy
     }
   }

   foo(false); // TS error.

   const lol: {} = false;
   foo(lol); // but this is allowed by TS.
   ```

As said, I think this should be _brief_, just to alert users to possibly-faulty common assumptions about the type system, and then punt deeper dive to https://github.com/microsoft/TypeScript/issues/60582#issuecomment-2547348195 or any other good references. The main goal is really to convey _think closely about your application before "fixing" a truthiness report, since the condition may actually be necessary, despite what the types appear to be indicate_.

### Affected URL(s)

https://main--typescript-eslint.netlify.app/rules/no-unnecessary-condition#limitations

### Additional Info

Relates to https://github.com/typescript-eslint/typescript-eslint/pull/10378, which may eventually change some of this behavior. But regardless of rule behavior change on our side, it's valuable for users to know that primitives may be lurking places they didn't expect.

**Comment by StyleShit** [comment]
already closed by #12111
